### ISSUE TITLE
Support serving `index.html` for StaticFiles

### ISSUE THREAD

**Issue by lovelydinosaur** [Kludex/starlette#146]
We should add an option for serving index.html files from StaticFiles for directory URLs so that you can use it for serving websites. We could also support serving any 404.html file in the root of the directory for 404 responses. (Ie. Mirror the behaviour of GitHub pages)

**Comment by jessebraham** [comment]
With the addition of #198 and the ability to provide a `template_directory` to the application, do we still want this feature? Seems like duplicated functionality to me but I'm sure there are use cases I haven't considered.

**Comment by jessebraham** [comment]
Almost done with this one. I'm assuming we want a feature switch in the `StaticFiles` constructor for this? Something to the effect of:

```python
def __init__(
    self,
    *,
    directory: str,
    check_dir: bool = True,
    serve_html: bool = False
) -> None
```

(I'm terrible at naming things ;) )
Just wanted to double check before making API changes all willy-nilly.

**Comment by lovelydinosaur** [comment]
Exactly yes.
I guess a long-winded `serve_index_html` would be okay here.

**Comment by kkinder** [comment]
Another possibility might be to patch [WhiteNoise](http://whitenoise.evans.io/en/stable/) to support ASGI and use WhiteNoise for this kind of use case.

**Comment by lovelydinosaur** [comment]
Fairly sure there’s an open ticket on WhiteNoise for ASGI support, yup.

**Comment by dgrahn** [comment]
Commenting here for future reference. Here's how you do it.

    StaticFiles(directory=..., html=True)

**Comment by lovelydinosaur** [comment]
Thanks @dgrahn, yup! Just released with 0.12.0.
